File: app/src/main/java/com/example/islandlyrics/AppLogger.java
package com.example.islandlyrics;

import android.os.Handler;
import android.os.Looper;
import androidx.lifecycle.MutableLiveData;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class AppLogger {
    private static AppLogger sInstance;
    private final MutableLiveData<String> logs = new MutableLiveData<>("");
    private final StringBuilder logBuffer = new StringBuilder();
    private final SimpleDateFormat dateFormat = new SimpleDateFormat("HH:mm:ss", Locale.US);
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    private AppLogger() {}

    public static synchronized AppLogger getInstance() {
        if (sInstance == null) {
            sInstance = new AppLogger();
        }
        return sInstance;
    }

    public MutableLiveData<String> getLogs() {
        return logs;
    }

    public void log(String tag, String message) {
        String timestamp = dateFormat.format(new Date());
        String logLine = String.format("[%s] [%s] %s\n", timestamp, tag, message);

        // Ensure UI updates happen on Main Thread
        mainHandler.post(() -> {
            logBuffer.append(logLine);
            // Keep buffer size for Deep Trace (approx last 100+ lines / 12KB)
            if (logBuffer.length() > 12000) {
                logBuffer.delete(0, logBuffer.indexOf("\n", 4000) + 1);
            }
            logs.setValue(logBuffer.toString());
        });
    }
}

File: app/src/main/java/com/example/islandlyrics/AppLogger.kt
package com.example.islandlyrics

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.lifecycle.MutableLiveData
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class AppLogger private constructor() {

    // Simplified: No buffering, no LiveData
    // val logs = MutableLiveData("") 
    // private val logBuffer = StringBuilder()
    
    // Check Debug Flag (Static build config)
    private val isDebug = BuildConfig.DEBUG
    
    // Runtime Logging Flag (Default to Debug status, but can be enabled in Release)
    var isLogEnabled = isDebug

    fun enableLogging(enable: Boolean) {
        isLogEnabled = enable
        if (enable) {
            i("AppLogger", "Logging Enabled by User")
        }
    }

    // DEBUG: Only if Enabled
    fun d(tag: String, message: String) {
        if (!isLogEnabled) return
        Log.d(tag, message)
    }

    // INFO: Only if Enabled (Release builds default to OFF)
    fun i(tag: String, message: String) {
        if (!isLogEnabled) return
        Log.i(tag, message)
    }

    // ERROR: Always to Logcat (for Crashlytics)
    fun e(tag: String, message: String) {
        e(tag, message, null)
    }

    fun e(tag: String, message: String, tr: Throwable?) {
        // Always log error to system for crash reporting
        if (tr != null) {
            Log.e(tag, message, tr)
        } else {
            Log.e(tag, message)
        }
    }

    // Legacy support: map 'log' to 'd' (Debug) or 'i' (Info) depending on importance
    fun log(tag: String, message: String) {
        d(tag, message)
    }
    
    // Legacy: No-op for buffer
    private fun appendToBuffer(level: String, tag: String, message: String) {
        // No-op
    }

    companion object {
        private var instance: AppLogger? = null

        @Synchronized
        fun getInstance(): AppLogger {
            if (instance == null) {
                instance = AppLogger()
            }
            return instance!!
        }
    }
}
